In [1]:
import pandas as pd
import numpy as np

In [45]:
df_or = pd.read_csv('green_tripdata_2015-01.csv')

In [274]:
df = df_or

In [275]:
df.tail()

VendorID lpep_pickup_datetime Lpep_dropoff_datetime  \
1048570         1     22/01/2015 14:56      22/01/2015 15:27   
1048571         1     22/01/2015 14:55      22/01/2015 15:13   
1048572         1     22/01/2015 14:55      22/01/2015 15:01   
1048573         1     22/01/2015 14:53      22/01/2015 15:01   
1048574         1     22/01/2015 14:52      22/01/2015 15:05   

        Store_and_fwd_flag  RateCodeID  Pickup_longitude  Pickup_latitude  \
1048570                  N           1        -73.959213        40.809345   
1048571                  N           1        -73.897095        40.675816   
1048572                  N           1        -73.874397        40.829514   
1048573                  N           1        -73.954147        40.787342   
1048574                  N           1        -73.963608        40.672054   

         Dropoff_longitude  Dropoff_latitude  Passenger_count     ...      \
1048570         -73.954941         40.765453                1     ...       
1048571         -73.937881         40.666531                1     ...       
1048572         -73.860619         40.837070                1     ...       
1048573         -73.971451         40.790321                2     ...       
1048574         -73.995300         40.679649                4     ...       

         Fare_amount  Extra  MTA_tax  Tip_amount  Tolls_amount  Ehail_fee  \
1048570         20.5    0.0      0.5        4.20           0.0        NaN   
1048571         14.0    0.0      0.5        0.00           0.0        NaN   
1048572          6.0    0.0      0.5        0.00           0.0        NaN   
1048573          7.0    0.0      0.5        1.55           0.0        NaN   
1048574         10.5    0.0      0.5        2.80           0.0        NaN   

         improvement_surcharge  Total_amount  Payment_type  Trip_type   
1048570                    0.0         25.20             1         1.0  
1048571                    0.3         14.80             4         1.0  
1048572                    0.3          6.80             2         1.0  
1048573                    0.3          9.35             1         1.0  
1048574                    0.3         14.10             1         1.0  

[5 rows x 21 columns]

In [276]:
df.shape

(1048575, 21)

### To delete: 
All rows in columns Pickup_longitude Pickup_latitude  Dropoff_longitude Dropoff_latitude with value 0.

In [277]:
df = df.drop(['Ehail_fee'],axis=1)

In [278]:
df.shape

(1048575, 20)

In [279]:
col = df.columns.tolist()

In [280]:
for column in col:
  print("{}: {}".format(column, df[column].isnull().sum()))

VendorID: 0
lpep_pickup_datetime: 0
Lpep_dropoff_datetime: 0
Store_and_fwd_flag: 0
RateCodeID: 0
Pickup_longitude: 0
Pickup_latitude: 0
Dropoff_longitude: 0
Dropoff_latitude: 0
Passenger_count: 0
Trip_distance: 0
Fare_amount: 0
Extra: 0
MTA_tax: 0
Tip_amount: 0
Tolls_amount: 0
improvement_surcharge: 0
Total_amount: 0
Payment_type: 0
Trip_type : 5


In [281]:
col = df.columns.tolist()

for column in col:
  print("{}: {}".format(column, df[column].unique()))

VendorID: [2 1]
lpep_pickup_datetime: ['01/01/2015 00:34' '01/01/2015 00:35' '01/01/2015 00:36' ...
 '22/01/2015 16:21' '22/01/2015 16:19' '22/01/2015 16:22']
Lpep_dropoff_datetime: ['01/01/2015 00:38' '01/01/2015 00:47' '01/01/2015 01:09' ...
 '22/01/2015 16:28' '22/01/2015 16:21' '22/01/2015 16:23']
Store_and_fwd_flag: ['N' 'Y']
RateCodeID: [ 1  5  2  3  4  6 99]
Pickup_longitude: [-73.92259216 -73.95275116 -73.84300995 ... -73.78638458 -73.85324097
 -74.0144043 ]
Pickup_latitude: [40.75452805 40.67771149 40.71905518 ... 40.88448334 40.63314438
 40.6570282 ]
Dropoff_longitude: [-73.91363525 -73.98152924 -73.84658051 ... -73.74492645 -73.65460968
 -74.18148804]
Dropoff_latitude: [40.765522   40.65897751 40.71156693 ... 40.59561157 40.88410568
 40.60508347]
Passenger_count: [1 4 5 2 6 3 8 0 7 9]
Trip_distance: [ 0.88  3.08  0.9  ... 26.27 22.39 25.62]
Fare_amount: [ 5.0000e+00  1.2000e+01  2.4500e+01  6.5000e+00  2.2000e+01  1.5000e+01
  7.0000e+00  8.0000e+00  1.6500e+01  3.9000e+01  

In [282]:
df.dtypes

VendorID                   int64
lpep_pickup_datetime      object
Lpep_dropoff_datetime     object
Store_and_fwd_flag        object
RateCodeID                 int64
Pickup_longitude         float64
Pickup_latitude          float64
Dropoff_longitude        float64
Dropoff_latitude         float64
Passenger_count            int64
Trip_distance            float64
Fare_amount              float64
Extra                    float64
MTA_tax                  float64
Tip_amount               float64
Tolls_amount             float64
improvement_surcharge    float64
Total_amount             float64
Payment_type               int64
Trip_type                float64
dtype: object

In [283]:
df.dropna(inplace = True)

## change types:
lpep_pickup_datetime Lpep_dropoff_datetime Trip_type 

## verify valids quantity of each column
No negatives:
Trip_distance Fare_amount Total_amount

In [284]:
df['lpep_pickup_datetime'] = pd.to_datetime(df['lpep_pickup_datetime'], format = '%d/%m/%Y %H:%M')

In [285]:
df['Lpep_dropoff_datetime'] = pd.to_datetime(df['Lpep_dropoff_datetime'], format = '%d/%m/%Y %H:%M')

In [286]:
df['Trip_type ']= df['Trip_type '].astype(np.int64)

In [287]:
df.dtypes

VendorID                          int64
lpep_pickup_datetime     datetime64[ns]
Lpep_dropoff_datetime    datetime64[ns]
Store_and_fwd_flag               object
RateCodeID                        int64
Pickup_longitude                float64
Pickup_latitude                 float64
Dropoff_longitude               float64
Dropoff_latitude                float64
Passenger_count                   int64
Trip_distance                   float64
Fare_amount                     float64
Extra                           float64
MTA_tax                         float64
Tip_amount                      float64
Tolls_amount                    float64
improvement_surcharge           float64
Total_amount                    float64
Payment_type                      int64
Trip_type                         int64
dtype: object

In [288]:
df['Trip_distance'].describe().round(2)

count    1048570.00
mean           2.78
std            2.95
min            0.00
25%            1.01
50%            1.83
75%            3.54
max          830.40
Name: Trip_distance, dtype: float64

In [289]:
df['Fare_amount'].describe().round(2)

count    1048570.00
mean          11.73
std            9.55
min         -400.00
25%            6.00
50%            9.00
75%           14.50
max          800.00
Name: Fare_amount, dtype: float64

In [290]:
df = df.loc[df['Fare_amount'] >= 0]

In [291]:
df['Total_amount'].describe().round(2)

count    1047038.00
mean          15.04
std          967.53
min            0.00
25%            7.80
50%           10.80
75%           17.00
max       989970.39
Name: Total_amount, dtype: float64

In [292]:
df = df.loc[df['Total_amount'] >= 0]

In [293]:
df.shape

(1047038, 20)

In [294]:
df.duplicated().sum()

117

In [295]:
df.drop_duplicates(inplace = True)

In [297]:
(df.describe()).round(2)

VendorID  RateCodeID  Pickup_longitude  Pickup_latitude  \
count  1046921.00  1046921.00        1046921.00       1046921.00   
mean         1.78        1.10            -73.82            40.69   
std          0.42        0.61              2.90             1.60   
min          1.00        1.00            -75.35             0.00   
25%          2.00        1.00            -73.96            40.70   
50%          2.00        1.00            -73.95            40.75   
75%          2.00        1.00            -73.92            40.80   
max          2.00       99.00              0.00            42.92   

       Dropoff_longitude  Dropoff_latitude  Passenger_count  Trip_distance  \
count         1046921.00        1046921.00       1046921.00     1046921.00   
mean              -73.82             40.68             1.39           2.79   
std                 2.95              1.63             1.08           2.95   
min               -75.44              0.00             0.00           0.00   
25%               -73.97             40.70             1.00           1.01   
50%               -73.95             40.75             1.00           1.83   
75%               -73.91             40.79             1.00           3.55   
max                 0.00             42.92             9.00         830.40   

       Fare_amount       Extra     MTA_tax  Tip_amount  Tolls_amount  \
count   1046921.00  1046921.00  1046921.00  1046921.00    1046921.00   
mean         11.76        0.34        0.49        1.12          0.10   
std           9.49        0.36        0.08        3.81          1.17   
min           0.00        0.00        0.00        0.00          0.00   
25%           6.00        0.00        0.50        0.00          0.00   
50%           9.00        0.50        0.50        0.00          0.00   
75%          14.50        0.50        0.50        1.80          0.00   
max         800.00        1.00        0.80      490.00        869.88   

       improvement_surcharge  Total_amount  Payment_type  Trip_type   
count             1046921.00    1046921.00    1046921.00  1046921.00  
mean                    0.28         15.04          1.56        1.02  
std                     0.07        967.58          0.52        0.15  
min                     0.00          0.00          1.00        1.00  
25%                     0.30          7.80          1.00        1.00  
50%                     0.30         10.80          2.00        1.00  
75%                     0.30         17.00          2.00        1.00  
max                     0.30     989970.39          5.00        2.00

In [298]:
df = df.loc[df['Pickup_longitude'] != 0]

In [299]:
df = df.loc[df['Pickup_latitude'] != 0]

In [300]:
df = df.loc[df['Dropoff_longitude'] != 0]

In [301]:
df = df.loc[df['Dropoff_latitude'] != 0]

In [302]:
df['RateCodeID'].unique()

array([ 1,  5,  2,  3,  4,  6, 99], dtype=int64)

In [303]:
df['RateCodeID'].value_counts()

1     1016585
5       23513
2        2862
3         715
4         348
6          20
99          1
Name: RateCodeID, dtype: int64

In [304]:
df = df.loc[df['RateCodeID'] != 99]
df.shape

(1044043, 20)

In [305]:
df['Total_amount'].describe().round(2)

count    1044043.00
mean          15.02
std          968.91
min            0.00
25%            7.80
50%           10.80
75%           17.00
max       989970.39
Name: Total_amount, dtype: float64

In [306]:
df = df.loc[df['Total_amount'] >= 0 ]
df = df.loc[df['Total_amount'] <= 1000]
df.shape 

(1044042, 20)

In [307]:
df['Total_amount'].describe().round(2)

count    1044042.00
mean          14.07
std           10.95
min            0.00
25%            7.80
50%           10.80
75%           17.00
max          874.18
Name: Total_amount, dtype: float64

In [308]:
df.describe().round(2)

VendorID  RateCodeID  Pickup_longitude  Pickup_latitude  \
count  1044042.00   1044042.0        1044042.00       1044042.00   
mean         1.78         1.1            -73.94            40.75   
std          0.41         0.6              0.04             0.06   
min          1.00         1.0            -75.35            39.95   
25%          2.00         1.0            -73.96            40.70   
50%          2.00         1.0            -73.95            40.75   
75%          2.00         1.0            -73.92            40.80   
max          2.00         6.0            -70.96            42.92   

       Dropoff_longitude  Dropoff_latitude  Passenger_count  Trip_distance  \
count         1044042.00        1044042.00       1044042.00     1044042.00   
mean              -73.94             40.75             1.39           2.79   
std                 0.05              0.06             1.08           2.95   
min               -75.44             39.36             0.00           0.00   
25%               -73.97             40.70             1.00           1.01   
50%               -73.95             40.75             1.00           1.83   
75%               -73.91             40.79             1.00           3.55   
max               -70.94             42.92             9.00         830.40   

       Fare_amount       Extra     MTA_tax  Tip_amount  Tolls_amount  \
count   1044042.00  1044042.00  1044042.00  1044042.00    1044042.00   
mean         11.76        0.34        0.49        1.11          0.10   
std           9.40        0.36        0.08        3.15          1.17   
min           0.00        0.00        0.00        0.00          0.00   
25%           6.00        0.00        0.50        0.00          0.00   
50%           9.00        0.50        0.50        0.00          0.00   
75%          14.50        0.50        0.50        1.80          0.00   
max         800.00        1.00        0.50      490.00        869.88   

       improvement_surcharge  Total_amount  Payment_type  Trip_type   
count             1044042.00    1044042.00    1044042.00  1044042.00  
mean                    0.28         14.07          1.56        1.02  
std                     0.07         10.95          0.52        0.15  
min                     0.00          0.00          1.00        1.00  
25%                     0.30          7.80          1.00        1.00  
50%                     0.30         10.80          2.00        1.00  
75%                     0.30         17.00          2.00        1.00  
max                     0.30        874.18          5.00        2.00

# Then we will split the dataset
The principal reason is because it's a good pratice dont work with all the samples, so for this reason and also because we hace a very large dataset.  

     total: 1044042  
 
 * 80 % train: 835233   
 * 20 % test: 208809
 
 Optionally:
 * 10 % validation

In [310]:
df.shape

(1044042, 20)

In [311]:
from random import sample
lil_df = df.loc[np.random.choice(df.index, 835233, replace = False)]
lil_df.shape

(835233, 20)

In [312]:
lil_df = lil_df.sort_values(by = 'lpep_pickup_datetime')

In [313]:
lil_df.reset_index(drop = True, inplace = True)

In [314]:
lil_df.head()

VendorID lpep_pickup_datetime Lpep_dropoff_datetime Store_and_fwd_flag  \
0         2           2015-01-01   2015-01-01 00:06:00                  N   
1         2           2015-01-01   2015-01-01 00:11:00                  N   
2         1           2015-01-01   2015-01-01 00:16:00                  N   
3         2           2015-01-01   2015-01-01 00:11:00                  N   
4         2           2015-01-01   2015-01-01 00:23:00                  N   

   RateCodeID  Pickup_longitude  Pickup_latitude  Dropoff_longitude  \
0           1        -73.961418        40.714340         -73.961357   
1           1        -73.890915        40.746918         -73.868408   
2           1        -73.951454        40.824905         -73.895561   
3           1        -73.990395        40.686752         -73.958351   
4           1        -73.952347        40.663719         -74.017433   

   Dropoff_latitude  Passenger_count  Trip_distance  Fare_amount  Extra  \
0         40.719212                1           0.74          6.0    0.5   
1         40.739296                1           1.67          8.5    0.5   
2         40.869946                1           5.70         19.0    0.5   
3         40.687611                1           1.88          9.5    0.5   
4         40.641983                1           6.01         21.0    0.5   

   MTA_tax  Tip_amount  Tolls_amount  improvement_surcharge  Total_amount  \
0      0.5        1.95           0.0                    0.0          8.95   
1      0.5        0.00           0.0                    0.3          9.80   
2      0.5        0.00           0.0                    0.3         20.30   
3      0.5        3.00           0.0                    0.3         13.80   
4      0.5        4.30           0.0                    0.3         26.60   

   Payment_type  Trip_type   
0             1           1  
1             2           1  
2             2           1  
3             1           1  
4             1           1

In [315]:
lil_df.duplicated().sum()

0

In [316]:
df_clean = lil_df

In [317]:
df_clean.describe()

VendorID     RateCodeID  Pickup_longitude  Pickup_latitude  \
count  835233.000000  835233.000000     835233.000000    835233.000000   
mean        1.779318       1.095375        -73.935918        40.749812   
std         0.414707       0.600378          0.041597         0.056858   
min         1.000000       1.000000        -75.351311        39.947926   
25%         2.000000       1.000000        -73.959366        40.700478   
50%         2.000000       1.000000        -73.945526        40.746872   
75%         2.000000       1.000000        -73.917969        40.802441   
max         2.000000       6.000000        -71.106323        42.918247   

       Dropoff_longitude  Dropoff_latitude  Passenger_count  Trip_distance  \
count      835233.000000     835233.000000    835233.000000  835233.000000   
mean          -73.935507         40.747841         1.389661       2.789763   
std             0.049663          0.057427         1.078698       2.985592   
min           -75.439667         39.357143         0.000000       0.000000   
25%           -73.968094         40.702679         1.000000       1.010000   
50%           -73.945648         40.747932         1.000000       1.830000   
75%           -73.912132         40.790722         1.000000       3.550000   
max           -71.089050         42.918175         9.000000     830.400000   

         Fare_amount          Extra        MTA_tax     Tip_amount  \
count  835233.000000  835233.000000  835233.000000  835233.000000   
mean       11.756595       0.343916       0.488236       1.106440   
std         9.425217       0.359569       0.075788       3.201608   
min         0.000000       0.000000       0.000000       0.000000   
25%         6.000000       0.000000       0.500000       0.000000   
50%         9.000000       0.500000       0.500000       0.000000   
75%        14.500000       0.500000       0.500000       1.800000   
max       800.000000       1.000000       0.500000     490.000000   

        Tolls_amount  improvement_surcharge   Total_amount   Payment_type  \
count  835233.000000          835233.000000  835233.000000  835233.000000   
mean        0.095862               0.281035      14.072366       1.558238   
std         1.250181               0.073005      10.996286       0.517392   
min         0.000000               0.000000       0.000000       1.000000   
25%         0.000000               0.300000       7.800000       1.000000   
50%         0.000000               0.300000      10.800000       2.000000   
75%         0.000000               0.300000      17.000000       2.000000   
max       869.880000               0.300000     874.180000       5.000000   

          Trip_type   
count  835233.000000  
mean        1.022446  
std         0.148130  
min         1.000000  
25%         1.000000  
50%         1.000000  
75%         1.000000  
max         2.000000

In [321]:
df_clean.dtypes

VendorID                          int64
lpep_pickup_datetime     datetime64[ns]
Lpep_dropoff_datetime    datetime64[ns]
Store_and_fwd_flag               object
RateCodeID                        int64
Pickup_longitude                float64
Pickup_latitude                 float64
Dropoff_longitude               float64
Dropoff_latitude                float64
Passenger_count                   int64
Trip_distance                   float64
Fare_amount                     float64
Extra                           float64
MTA_tax                         float64
Tip_amount                      float64
Tolls_amount                    float64
improvement_surcharge           float64
Total_amount                    float64
Payment_type                      int64
Trip_type                         int64
dtype: object

(835233, 20)

In [ ]:
data_filter = dfg.loc[dfg['Pickup_longitude'] != (0)]
#df_clean = points.loc[points['Pickup_longitude'] != str(0)]

In [ ]:
data_filter = data_filter.loc[data_filter['Pickup_latitude'] != (0)]

In [ ]:
data_filter = data_filter.loc[data_filter['Dropoff_longitude'] != (0)]

In [ ]:
data_filter.shape

In [ ]:
x_coords = data_filter['Dropoff_longitude'].astype(float)
y_coords = data_filter['Pickup_latitude'].astype(float)

In [ ]:
col = data_filter.columns.tolist()

In [ ]:
for column in col:
  print("{}: {}".format(columns, datafilter[column].isnull().sum()))

In [ ]:
dfg.shape

In [ ]:
dfg.head()

In [ ]:
dfn = dfg.drop(dfg.columns[[16]], axis = 'columns')

In [ ]:
dfn = dfn.dropna()

In [ ]:
dfn.head()

In [ ]:
col = dfn.columns.tolist()
for column in col:
  print("{}: {}".format(column, dfn[column].isnull().sum()))

In [ ]:
dfn.shape

In [ ]:
dfg.describe()

In [ ]:
dfg.info()

In [ ]:
from bokeh.plotting import figure, save
p = figure(title="My first plot!")
p

In [ ]:
x_coords = [0,1,2,3,4]
y_coords = [5,4,1,2,0]

In [ ]:
p.circle(x=x_coords, y = y_coords, size = 10, color = "red")

In [ ]:
#C:\Users\alter\Desktop
outputfp = r"/Users/alter/Desktop/points.htmls"

In [ ]:
save(obj = p, filename = outputfp)

In [ ]:
import geopandas as gpd

In [ ]:
points_fp = r"/Users/alter/Documents/Barcelona/La Salle/Asignaturas/Data Mining/JupyterGH/green_tripdata_2015-02.csv"

# Read the data
points = gpd.read_file(points_fp)

In [ ]:
points.head()

In [ ]:
p = figure(title="A map of address points from a Shapefile")